### ISSUE TITLE
write_dataset returns nothing

### ISSUE THREAD

**Issue by r2evans** [apache/arrow#40723]
### Describe the bug, including details regarding any error messages, version, and platform.

The docs say

```
write_dataset              package:arrow               R Documentation

Write a dataset

Description:

     This function allows you to write a dataset. By writing to more
     efficient binary storage formats, and by specifying relevant
     partitioning, you can make it much faster to read and query.

Usage:

     write_dataset(
       dataset,
       path,
       format = c("parquet", "feather", "arrow", "ipc", "csv", "tsv", "txt", "text"),
       partitioning = dplyr::group_vars(dataset),
       basename_template = paste0("part-{i}.", as.character(format)),
       hive_style = TRUE,
       existing_data_behavior = c("overwrite", "error", "delete_matching"),
       max_partitions = 1024L,
       max_open_files = 900L,
       max_rows_per_file = 0L,
       min_rows_per_group = 0L,
       max_rows_per_group = bitwShiftL(1, 20),
       ...
     )

Value:

     The input ‘dataset’, invisibly
```

but when I write datasets, I get nothing in return:

```r
res <- arrow::write_dataset(mtcars, "mt")
res
# NULL
```

On Ubuntu-23.10, R-4-3.2, arrow_15.0.1.

### Component(s)

R

**Comment by r2evans** [comment]
Part of me says I don't _need_ the data, I'd rather have all of the filenames created ... but that's more of a feature-request than a bug-report.

```r
res <- arrow::write_dataset(mtcars, "mt", partitioning = c("cyl", "gear"))
res
# NULL
list.files("mt", recursive = TRUE, full.names = TRUE)
# [1] "mt/cyl=4/gear=3/part-0.parquet" "mt/cyl=4/gear=4/part-0.parquet" "mt/cyl=4/gear=5/part-0.parquet" "mt/cyl=6/gear=3/part-0.parquet"
# [5] "mt/cyl=6/gear=4/part-0.parquet" "mt/cyl=6/gear=5/part-0.parquet" "mt/cyl=8/gear=3/part-0.parquet" "mt/cyl=8/gear=5/part-0.parquet"
```

That is, I don't want to need to do `list.files(..)`, since `arrow` knows internally what files were created. A recursive search can get a little annoying if there are many existing files.

**Comment by thisisnic** [comment]
Thanks for reporting this @r2evans.  

I can confirm that I can reproduce this - nothing is returned.

Thinking through the various options of what would make sense to return, I had a few thoughts.

Looking at other examples, such as `DBI::dbWriteTable()` (as sometimes we think of working with datasets as analogous to working with a database connection), that function just returns `TRUE` invisibly to indicate success.

However, looking at the [tidyverse design guide](https://design.tidyverse.org/out-invisible.html), a principle referenced there is:

> If a function is called primarily for its side-effects, it should invisibly return a useful output. If there’s no obvious output, return the first argument. This makes it possible to use the function with in a pipeline.

I'm not sure how common it is to use `write_dataset()` mid-pipeline, though it's feasible, so that could make sense too.

I can see why you might want to have the list of files created, but I'm a bit hesitant to diverge from the behaviour of existing similar APIs, and this might be better thought through as a separate feature request.

Curious to see what @amoeba thinks here?

**Comment by r2evans** [comment]
I agree that returning something useful is a good thing. Often, yes, mid-pipeline functions pass the data through (often invisibly), but I agree that I don't see this being a mid-pipeline kind of operation.

My thought is to return the filenames created. While we are calling this function primarily for its side-effect, what we don't know is exactly what filenames will be created; the only time we "know" is when we're writing completely new data (partitioning subdirs do not exist or are known to be empty). If that is ever important (I have an internal use-case where it is), then we have to _infer_ what files were created, something that can be ambiguous in situations where there are pre-existing files.

Do you think returning a `character` vector with the final filenames is both meaningful and feasible?

Another argument for this: _if_ this returns files and it is being used mid-pipe where data would be more useful, there is an easy workaround:

```r
# dplyr pipe
dat %>%
  mutate(...) %>%
  { write_dataset(., ...); .; } %>%
  summarize(...)
# native pipe
dat |>
  transform(..) |>
  (\(.x) { write_dataset(.x, ...); .x; })() |>
  aggregate(..., data = _)
```

But if the function instead returns data, there is no unambiguous way to immediately know the filenames created.

**Comment by amoeba** [comment]
I wonder if we could hook things up so it returns a Dataset object for the _serialized_ Dataset, instead of the input dataset. I too am not sure about the use case of putting `write_dataset` mid-pipeline.  

Beyond the above use case, it would be a nice time-saver for cases where you want to call `open_dataset` on the newly-serialized dataset but where it might take some time due to dataset discovery (particularly on cloud storage). I think when `write_dataset` returns it may already know enough to create a Dataset object and to skip discovery (schema, fragments, partitioning).

Getting the files out of `write_dataset` could then be done like this:

```r
ds %>% 
  write_dataset("outdir") %>%
   .$files
```

and we wouldn't be limiting the information you could get out of `write_dataset` to just the files.

**Comment by r2evans** [comment]
That's a good suggestion, I like that option. Thanks @amoeba!

**Comment by amoeba** [comment]
Right. I'd only want us to do what I suggest if it was nearly free (not sure if can be made free). Adding an argument to `write_dataset` that changes its behavior, while useful, doesn't feel quite right.

For uniformity's sake, I'm inclined to follow @thisisnic's lead and follow the tidyverse principle here (return the input dataset). It doesn't improve things for your use case but it does fix the original bug.

**Comment by r2evans** [comment]
I understand preferring simplicity, but ... I cannot think of a situation where I would use `write_dataset` in the middle of a data pipe and expect the data to continue. The premise of `write_dataset` is on its _side-effect_, not on its return value. (I don't understand it for `readr::write_csv`, either.) I suggest that the strength in returning the filenames-created is that the underlying code (somewhere) already knows what the filenames are, so there should be no need to scan for them. Scanning for them is exacerbated when we use partitioning into existing path with existing partitions. Whether we delete existing files, use a different `basename_template=`, or whatever, when partitions pre-exist, there are no "_inexpensive_" ways to determine if my recent call to `write_dataset` is what created the files I see. (I characterize "scanning all files" as relatively expensive. I do much work on an HPC using a filesystem that has, at times, had a significant 5-10 second _lag_, but that is just anecdotal for why I suggest that forcing a scan of the filesystem should not be necessary.)

Think of this:

```r
arrow::write_dataset(mtcars[1:3,], td, partition = "cyl")
list.files(td, recursive = TRUE)
# [1] "cyl=4/part-0.parquet" "cyl=6/part-0.parquet"
### pause ...
arrow::write_dataset(mtcars[4:5,], td, partition = "cyl")
list.files(td, recursive = TRUE)
# [1] "cyl=4/part-0.parquet" "cyl=6/part-0.parquet" "cyl=8/part-0.parquet"
```

Which of those files were created on the second write? If you know the data, then you'll know that `cyl=4` is not present in rows 4-5, so if we disect the data we can figure that out, but consider millions of rows and this becomes a bit more onerous.

Yes, we _can_ do a scan of the directory structure, and we'll see which files were created in the second batch by comparing the `ctime` values:

```r
file.info(list.files(td, recursive = TRUE, full.names = TRUE), extra_cols = FALSE)
#                                                                 size isdir mode               mtime               ctime               atime
# /home/r2/tmp/Rtmp1ZaQHP/file158b58c3a3b4.d/cyl=4/part-0.parquet 4579 FALSE  664 2024-03-28 19:57:03 2024-03-28 19:57:03 2024-03-28 19:57:03
# /home/r2/tmp/Rtmp1ZaQHP/file158b58c3a3b4.d/cyl=6/part-0.parquet 4579 FALSE  664 2024-03-28 19:58:11 2024-03-28 19:58:11 2024-03-28 19:57:03
# /home/r2/tmp/Rtmp1ZaQHP/file158b58c3a3b4.d/cyl=8/part-0.parquet 4579 FALSE  664 2024-03-28 19:58:11 2024-03-28 19:58:11 2024-03-28 19:58:11
```

But do we really need to do that? The underlying code _knows_ (somewhere) that it only wrote to the `cyl=6` and `cyl=8` partitions.

Conversely, it returns filenames and we want it to do the "tidyverse principle" of returning data, we can _trivially_ get that with

```r
some %>%
  pipe() %>%
  { write_dataset(path=td); .; } %>%
  something_more()
```

Granted, if you are amenable to adding an option or two that facilitates either or both of those pathways, then it could be as simple as one of the following:

```r
some %>%
  pipe() %>%
  write_dataset(path=td) %>%
  something_more()
newfiles <- write_dataset(path=td, return_files=TRUE)
```

or

```r
some %>%
  pipe() %>%
  write_dataset(path=td, return_data=TRUE) %>%
  something_more()
```

or some variant of either/both of those.

**Comment by r2evans** [comment]
My suggestion is not without precedent, I believe. In the "tidy design principles", it does say

> `fs:file_copy(from, to)` returns `to`


Which is not too dissimilar from returning the filenames. I think comparing `write_parquet` to `write_csv` is perfectly legitimate, since the filename is unambiguous. However, since `write_dataset` can obviously be significantly different than what it was called with, I think there's precedent and value for that.

**Comment by thisisnic** [comment]
Agree with @amoeba that we'd only want to do this if it could be done for nearly free; with that in mind, if you'd like to submit a PR, @r2evans, I wouldn't be opposed to the change.

**Comment by r2evans** [comment]
I understand that you want to add zero or near-zero overhead, I completely agree. Does the underlying library not have the filenames it created readily available?

**Comment by r2evans** [comment]
Does the underlying `plan` or `ExecPlan_Write` have any attributes that can be directly used to know or smartly infer the filenames?

**Comment by thisisnic** [comment]
There hasn't been any further discussion, but you're still welcome to submit a PR and as long as there's no additional overhead, then it still sounds like a reasonable change to me.

**Comment by thisisnic** [comment]
It doesn't appear that there's any ongoing interest in this, so I'm closing this.
